### DISCUSSION TITLE
FlowJo workspace import

### DISCUSSION THREAD

**Discussion by schienstockd** [ImmuneDynamics/Spectre#78] [! News]
Hello,

You have a cool framework - I was just wondering whether I could import a FlowJo Workspace directly into spectre?

I have seen your tutorial:
https://wiki.centenary.org.au/display/SPECTRE/Exporting+data+from+FlowJo+for+analysis+in+Spectre

Could we utilise flowWorkspace somehow?
https://github.com/RGLab/flowWorkspace

Thanks!

Dominik

**Comment by tomashhurst** [comment]
Hi @schienstockd , sorry for the late reply! I have actually been thinking about this quite a bit. I am keen to get something like this working, and I'm sure it can be done, but we are a bit tied up with some other major developments at the moment. Are you using flowWorkspace at the moment?

**Comment by schienstockd** [comment]
Yes, I analyse my samples in R with flowWorkspace, openCyto and ggcyto. Most people in the lab use FlowJo for gating but some would like to use R for further analysis. It would be good to make this transition relatively painless.

How can I load data into spectre? Do I need to save the populations as CSV or could I also push a list of data.tables somehow?
Something like this?

```R
library(CytoML)
library(ggcyto)

# read FlowJo workspace
wsfile <- "path/to/workspace.wsp"
ws <- open_flowjo_xml(wsfile)

# get second group
gs <- flowjo_to_gatingset(ws, name = 2)

# get a list of data.tables
subsets <- c("T cells", "B cells", "DC")

wsSubsets <- list()
for (subset in subsets) {
  # fortify gating set
  attr(gs, "subset") <- subset
  wsSubsets[[subset]] <- fortify(gs)
}

# How to open that in spectre.. ?
```

**Comment by tomashhurst** [comment]
It should very much be possible to extract a list of `data.table`s, as you have indicated above. Spectre essentially just manages data as a large table of cells (rows) vs features/markers (columns). So you can pull the matrix of values for each file, and merge them together into a `data.table` and just proceed straight into any of the Spectre workflows. This would be a great addition, as it would be handy way of rapidly pulling in pre-labelled data via gating (rather than going to the effort of exporting the gates etc).

I'll try out this approach on Monday, as you've peaked my curiosity, but you might get an answer before that. What exactly does the `fortify` function gives you in your loop?

**Reply by schienstockd** [reply]
`fortify` is a function from the [ggcyto package](https://www.rdocumentation.org/packages/ggcyto/versions/1.0.5/topics/fortify.flowFrame). They implemented it primarily for plotting the data from a GatingSet with ggplot, I think.

It will create a `data.table` with the channel intensities from the FCS file and the associated pData.
I am not sure whether extra information from the FlowJo workspace will also be included in there, for example, when you assign specific treatments or timepoints to the individual tubes/wells.

**Comment by denvercal1234GitHub** [comment]
Would you mind helping me understand the advantage of importing the whole workspace versus what is being described here (https://wiki.centenary.org.au/display/SPECTRE/Exporting+data+from+FlowJo+for+analysis+in+Spectre)? I thought exporting gated population of interest from each sample as csv would give me more control how to process these data before downstream analysis? Thank you all!

**Reply by schienstockd** [reply]
The idea is to automate the export of gated populations from the workspace - not to import the whole workspace. The advantage would be that if you change the gating in your workspace you would only have to run this import script/function and have the changed populations in spectre without having to manually export them again from FlowJo.

It seems that you could pull them from the workspace, create a list of `data.table`s, then merge them into one big `data.table` and use the populations in spectre.

**Reply by SamGG** [reply]
The idea is very interesting. Making the process from workspace to spectre as direct as possible would attract a lot of analysts. And FlowJo is used a lot.

**Reply by tomashhurst** [reply]
It would potentially be advantageous when comparing gates and and clusters, and being able to select gated subsets in a flexible way during analysis is quite appealing -- rather than exporting specific gates etc and then going back and forth through that process.

**Comment by schienstockd** [comment]
Here is a naive start. Then you will have all populations in one `data.table`.

```R
library(CytoML)
library(ggcyto)
library(data.table)

# Load FlowJo workspace
load.flowjo.gs <- function(wsp.file, sample.group = 1) {
  # read FlowJo workspace
  ws <- open_flowjo_xml(wsp.file)
  
  # get sample group
  gs <- flowjo_to_gatingset(ws, name = sample.group)
  
  gs
}

# get gated populations from FlowJo workspace
gated.pops.to.data.table <- function(gs, gated.pops) {
  # get subsets
  subset.list <- lapply(gated.pops, function(x) {
    # fortify gating set
    attr(gs, "subset") <- x
    y <- fortify(gs)
    
    # add population name
    y[, ("FlowPop") := x]
    
    y
  })
  
  # bind 
  DT <- rbindlist(subset.list)
  
  # prepare data table
  DT[, c(".rownames") := NULL]
  setnames(DT, "name", "FileName")
  DT[, FileNo := .GRP, by = .(FileName)]
  
  DT
}

# load workspace
flowjo.gs <- load.flowjo.gs("path/to/workspace.wsp", sample.group = 2)

# get population table
gated.pops <- gated.pops.to.data.table(flowjo.gs, c("cDC1", "cDC2", "pDC"))
```

**Reply by tomashhurst** [reply]
Badass! I'm definitely giving this a shot later

**Reply by schienstockd** [reply]
You might have to checkout the reference for [`flowjo_to_gatingset`](https://rglab.github.io/CytoML/reference/flowjo_to_gatingset.html). There are a few options to specify the FCS path, transformation and compensation.
